plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java

```diff
@@ -0,0 +1,180 @@
+    protected final ThreadPool threadPool;
+    protected volatile Scheduler.Cancellable scheduledFuture;
+
+    @Inject
+    public QueryInsightsService(ThreadPool threadPool, @Nullable S store, @Nullable E exporter) {
+        this.threadPool = threadPool;
+        this.store = store;
+        this.exporter = exporter;
+    }
+
+    /**
+     * Ingest one record to the query insight store
+     *
+     * @param record the record to ingest
+     */
+    protected void ingestQueryData(R record) {
+        if (this.store != null) {
+            this.store.add(record);
+        }
+    }
+
+    /**
+     * Get all records that are in the query insight store,
+     * By default, return the records in sorted order.
+     *
+     * @return List of the records that are in the query insight store
+     * @throws IllegalArgumentException if query insight is disabled in the cluster
+     */
+    public List<R> getQueryData() throws IllegalArgumentException {
+        if (!enableCollect) {
+            throw new IllegalArgumentException("Cannot get query data when query insight feature is not enabled.");
+        }
+        clearOutdatedData();
+        List<R> queries = new ArrayList<>(store);
+        queries.sort(Collections.reverseOrder());
+        return queries;
+    }
+
+    /**
+     * Clear all outdated data in the store
+     */
+    public abstract void clearOutdatedData();
+
+    /**
+     * Restart the exporter with new config
+     */
+    public abstract void resetExporter(boolean enabled, QueryInsightsExporterType type, String identifier);
+
+    /**
+     * Clear all data in the store
+     */
+    public void clearAllData() {
+        store.clear();
+    }
+
+    public void setEnableCollect(boolean enableCollect) {
+        this.enableCollect = enableCollect;
+    }
+
+    public boolean getEnableCollect() {
+        return this.enableCollect;
+    }
+
+    public void setEnableExport(boolean enableExport) {
+        this.enableExport = enableExport;
+    }
+
+    public boolean getEnableExport() {
+        return this.enableExport;
+    }
+
+    /**
+     * Start the Query Insight Service.
+     */
+    @Override
+    protected void doStart() {
+        if (exporter != null && getEnableExport()) {
+            scheduledFuture = threadPool.scheduleWithFixedDelay(this::doExport, exportInterval, ThreadPool.Names.GENERIC);
+        }
+    }
+
+    /**
+     * Stop the Query Insight Service
+     */
+    @Override
+    protected void doStop() {
+        if (scheduledFuture != null) {
+            scheduledFuture.cancel();
+            if (exporter != null && getEnableExport()) {
+                doExport();
+            }
+        }
+    }
+
+    private void doExport() {
+        List<R> storedData = getQueryData();
+        try {
+            exporter.export(storedData);
+            log.debug(String.format(Locale.ROOT, "finish exporting query insight data to sink %s", storedData));
```

Comment:
The semantics of export method are async, so you log success even in case it does not finish.

Reply:
make sense, let me remove this debug log and rely only on the log from inside the async func.